package com.leetcode2.org.树状数组;

public class Solution1409 {
    public int[] processQueries(int[] queries, int m) {
        int n = queries.length;
        BIT bit = new BIT(m + n);

        int[] pos = new int[m + 1];
        for (int i = 1; i <= m; ++i) {
            pos[i] = n + i;
            bit.update(n + i, 1);
        }

        int[] ans = new int[n];
        for (int i = 0; i < n; ++i) {
            int cur = pos[queries[i]];
            bit.update(cur, -1);
            ans[i] = bit.query(cur);
            cur = n - i;
            pos[queries[i]] = cur;
            bit.update(cur, 1);
        }
        return ans;
    }
    static class BIT {
        int[] a;
        int n;

        public BIT(int n) {
            this.n = n;
            this.a = new int[n + 1];
        }

        public int query(int x) {
            int ret = 0;
            while (x != 0) {
                ret += a[x];
                x -= lowbit(x);
            }
            return ret;
        }

        public int update(int x, int dt) {
            while (x <= n) {
                a[x] += dt;
                x += lowbit(x);
            }
            return x;
        }

        public static int lowbit(int x) {
            return x & (-x);
        }


    }
}